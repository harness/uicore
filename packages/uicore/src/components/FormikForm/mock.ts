/*
 * Copyright 2024 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export const mockData = [
  {
    id: '1',
    name: 'Google Pixel 6 Pro',
    data: {
      color: 'Cloudy White',
      capacity: '128 GB'
    }
  },
  {
    id: '2',
    name: 'Apple iPhone 12 Mini, 256GB, Blue',
    data: null
  },
  {
    id: '3',
    name: 'Apple iPhone 12 Pro Max',
    data: {
      color: 'Cloudy White',
      'capacity GB': 512
    }
  },
  {
    id: '4',
    name: 'Apple iPhone 11, 64GB',
    data: {
      price: 389.99,
      color: 'Purple'
    }
  },
  {
    id: '5',
    name: 'Samsung Galaxy Z Fold2',
    data: {
      price: 689.99,
      color: 'Brown'
    }
  },
  {
    id: '6',
    name: 'Apple AirPods',
    data: {
      generation: '3rd',
      price: 120
    }
  },
  {
    id: '7',
    name: 'Apple MacBook Pro 16',
    data: {
      year: 2019,
      price: 1849.99,
      'CPU model': 'Intel Core i9',
      'Hard disk size': '1 TB'
    }
  },
  {
    id: '8',
    name: 'Apple Watch Series 8',
    data: {
      'Strap Colour': 'Elderberry',
      'Case Size': '41mm'
    }
  },
  {
    id: '9',
    name: 'Beats Studio3 Wireless',
    data: {
      Color: 'Red',
      Description: 'High-performance wireless noise cancelling headphones'
    }
  },
  {
    id: '10',
    name: 'Apple iPad Mini 5th Gen',
    data: {
      Capacity: '64 GB',
      'Screen size': 7.9
    }
  },
  {
    id: '11',
    name: 'Apple iPad Mini 5th Gen',
    data: {
      Capacity: '254 GB',
      'Screen size': 7.9
    }
  },
  {
    id: '12',
    name: 'Apple iPad Air',
    data: {
      Generation: '4th',
      Price: '419.99',
      Capacity: '64 GB'
    }
  },
  {
    id: '13',
    name: 'Apple iPad Air',
    data: {
      Generation: '4th',
      Price: '519.99',
      Capacity: '256 GB'
    }
  }
]
