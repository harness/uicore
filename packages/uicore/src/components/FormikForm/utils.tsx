/*
 * Copyright (c) Harness Inc.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { ReactNode, useContext } from 'react'
import { FormikContextType } from 'formik'

import { get } from 'lodash-es'
import { FormikTooltipContext } from './FormikTooltipContext'
import { HarnessDocTooltip } from '../../frameworks/Tooltip/Tooltip'
import { DataTooltipInterface } from '../../frameworks/Tooltip/types'

export interface FormikExtended<T> extends FormikContextType<T> {
  disabled?: boolean
  inline?: boolean
  formName: string
}

export interface FormikContextProps<T> {
  formik?: FormikExtended<T>
  optionalLabel?: string
  isOptional?: boolean // default to false
  autoComplete?: string
  tooltipProps?: DataTooltipInterface // todo mark it as mandatory
}

export const IsOptionLabel = '(optional)'

const isObject = (obj: any): boolean => obj !== null && typeof obj === 'object'

export const errorCheck = (name: string, formik?: FormikContextType<any>): boolean =>
  Boolean(
    (get(formik?.touched, name) || (formik?.submitCount && formik?.submitCount > 0)) &&
      get(formik?.errors, name) &&
      !isObject(get(formik?.errors, name))
  )

export const getFormFieldLabel = (
  label: ReactNode | string | undefined,
  fieldName: string,
  props: FormikContextProps<any>,
  css?: string
): ReactNode | string | undefined => {
  const optionalLabel = props.optionalLabel || IsOptionLabel
  const labelText = !props.isOptional ? label : `${label} ${optionalLabel}`
  if (!labelText) {
    return labelText
  }
  const tooltipContext = useContext(FormikTooltipContext)
  const dataTooltipId =
    props.tooltipProps?.dataTooltipId || (tooltipContext?.formName ? `${tooltipContext?.formName}_${fieldName}` : '')
  return <HarnessDocTooltip tooltipId={dataTooltipId} labelText={labelText} className={css || ''} />
}

export function escapeNewlines(input: string | number): string | number {
  if (typeof input === 'string') {
    return input.replace(/\n/g, '\\n').replace(/\r/g, '\\r')
  }
  return input
}
