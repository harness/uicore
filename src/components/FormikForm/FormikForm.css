/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

@define-mixin intent {
  &.bp3-intent-primary {
    --bp3-intent-color: var(--primary-7);
  }

  &.bp3-intent-success {
    --bp3-intent-color: var(--green-500);
  }

  &.bp3-intent-warning {
    --bp3-intent-color: var(--yellow-500);
  }

  &.bp3-intent-danger {
    --bp3-intent-color: var(--red-500);
  }
}

.main {
  :global {
    .bp3-form-group {
      @mixin intent;

      .bp3-form-helper-text {
        color: var(--bp3-intent-color, var(--grey-500)) !important;
      }
      .bp3-file-upload-input,
      .ColorPicker--picker {
        box-shadow: none !important;
        border: 1px solid var(--grey-200) !important;
      }

      .bp3-input {
        height: var(--spacing-8);
        border-radius: var(--spacing-2);
        border: 1px solid var(--bp3-intent-color, var(--grey-200));
        outline: none;
        box-shadow: none;

        &:hover {
          border-color: var(--primary-6);
          box-shadow: 0px 0px 1px rgba(40, 41, 61, 0.04), 0px 2px 4px rgba(96, 97, 112, 0.16);
        }

        &:focus {
          border-color: var(--primary-7);
        }

        &:disabled {
          background-color: var(--grey-50);
          color: var(--grey-400);
          border-color: var(--grey-200);
          box-shadow: none;
        }
      }
    }

    .bp3-label {
      color: var(--grey-600);
      font-size: var(--form-input-font-size);
      font-weight: 500;
    }

    .bp3-input.bp3-tag-input {
      overflow: auto;
      height: fit-content;
      max-height: 70px;
      &.bp3-disabled {
        background: var(--grey-50) !important;
      }
      &.bp3-active {
        --bp3-intent-color: var(--primary-5);
      }

      .bp3-tag-input-values {
        margin: 2px 7px 0 1px !important;
      }
    }

    .bp3-file-input {
      height: 38px;
    }

    textarea.bp3-input {
      min-height: var(--spacing-10);
    }

    .bp3-file-upload-input {
      height: 38px;
      border-radius: 5px;
      box-shadow: none;
      border: 1px solid #dddddd;
      line-height: 36px;
      &:after {
        line-height: 28px;
        border: 1px solid var(--grey-350);
        box-shadow: none;
        background-image: none;
        color: var(--grey-600);
        background-color: var(--white);
      }
    }
    .bp3-input-ghost {
      font-size: 14px;
      line-height: 22px;
      padding-top: 2px;
    }

    .bp3-tag-input .bp3-tag {
      height: 22px;
      line-height: 22px;
      font-size: 14px;
      margin-top: 3px;
    }
  }
}

.checkBoxDocTooltipLabel {
  display: inline-block !important;
}

.inlineRadioGroup {
  padding-top: var(--spacing-2) !important;
}
