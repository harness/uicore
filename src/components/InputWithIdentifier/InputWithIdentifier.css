.txtNameContainer {
  position: relative;

  .txtIdContainer {
    position: absolute;
    align-items: center;
    right: 0;
  }

  .idValue {
    max-width: 100px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    color: var(--blue-600);
  }
}
