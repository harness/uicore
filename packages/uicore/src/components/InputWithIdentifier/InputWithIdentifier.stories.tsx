/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { noop } from 'lodash-es'

import { InputWithIdentifier, InputWithIdentifierProps } from './InputWithIdentifier'
import { Formik, FormikForm } from '../FormikForm/FormikForm'
import { Container } from '../Container/Container'

export default {
  title: 'Form / InputWithIdentifier',
  component: InputWithIdentifier
} as Meta

export const Basic: Story<InputWithIdentifierProps> = args => (
  <Formik initialValues={{}} onSubmit={noop} formName="inputWithIdentifierForm">
    {formikProps => (
      <Container width={400}>
        <FormikForm>
          <InputWithIdentifier {...args} formik={formikProps} />
        </FormikForm>
      </Container>
    )}
  </Formik>
)

Basic.args = {}
