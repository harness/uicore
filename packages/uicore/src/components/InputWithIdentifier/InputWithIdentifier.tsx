/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useState, useContext } from 'react'
import { EditableText, Popover, PopoverInteractionKind } from '@blueprintjs/core'
import { get, isNil, set, cloneDeep } from 'lodash-es'
import cx from 'classnames'
import { FormikTooltipContext } from '../FormikForm/FormikTooltipContext'

import { FormInput, TextProps } from '../FormikForm/FormikForm'
import { Text } from '../Text/Text'
import { Icon } from '@harness/icons'
import { Layout } from '../../layouts/Layout'
import { Container } from '../Container/Container'
import textCss from '../TextInput/TextInput.css'
import css from './InputWithIdentifier.css'
import { FormError } from '../FormError/FormError'

export interface InputWithIdentifierProps {
  formik: any
  inputGroupProps?: Omit<TextProps, 'name' | 'label' | 'onChange'>
  /**
   * @default name
   */
  inputName?: string
  /**
   * @default identifier
   */
  idName?: string
  /**
   * @default Name
   */
  inputLabel?: string
  /**
   * @default ID
   */
  idLabel?: string
  /**
   * Should the user be allowed to edit the identifier?
   * @default true
   */
  isIdentifierEditable?: boolean
  /**
   * @default 128
   */
  maxInput?: number
  /**
   * @default true
   */
  useUnversialToolTipId?: boolean
  /**
   * callback function that gets triggered when
   * identifier changes, either directly or due to changes in name field
   */
  onIdentifierChangeCallback?: (identifier: string) => void
}

// https://harness.atlassian.net/wiki/spaces/CDNG/pages/736200458/Entity+Identifier
export function getIdentifierFromName(str: string): string {
  if (isNil(str)) return ''

  return str
    .trim()
    .replace(/^[0-9-$]*/, '') // remove starting digits, dashes and $
    .replace(/[^0-9a-zA-Z_$ ]/g, '') // remove special chars except _ and $
    .replace(/ +/g, '_') // replace spaces with _
}

export const InputWithIdentifier: React.FC<InputWithIdentifierProps> = props => {
  const {
    formik,
    inputLabel = 'Name',
    idLabel = 'Id',
    inputName = 'name',
    idName = 'identifier',
    inputGroupProps,
    isIdentifierEditable = true,
    maxInput = 128,
    useUnversialToolTipId = true,
    onIdentifierChangeCallback
  } = props
  const [editable, setEditable] = useState(false)
  const [userModifiedIdentifier, setUserModifiedIdentifier] = useState(false)
  const identifier = get(formik.values, idName) as string
  const [currentEditField, setCurrentEditField] = useState(inputLabel)
  const tooltipContext = useContext(FormikTooltipContext)
  const dataTooltipId = !useUnversialToolTipId
    ? tooltipContext?.formName
      ? `${tooltipContext?.formName}_${idName}`
      : ''
    : 'universalInputNameWithId'

  return (
    <div className={cx(css.txtNameContainer, textCss.main)}>
      <Layout.Horizontal className={css.txtIdContainer} spacing="xsmall">
        <Text
          tooltipProps={{
            dataTooltipId
          }}>
          {idLabel}
        </Text>
        <Text>:</Text>
        <Popover
          content={<Container padding="small">{identifier}</Container>}
          interactionKind={PopoverInteractionKind.HOVER}
          hoverOpenDelay={500}
          disabled={!identifier || identifier.length < 10}>
          <EditableText
            maxLength={maxInput}
            disabled={!isIdentifierEditable}
            placeholder=""
            minWidth={0}
            value={identifier}
            isEditing={editable}
            onConfirm={() => setEditable(false)}
            onCancel={() => setEditable(false)}
            onEdit={() => setEditable(true)}
            onChange={value => {
              const identifierFromName = getIdentifierFromName(value)
              setCurrentEditField(idName)
              formik.setFieldValue(idName, identifierFromName)
              setUserModifiedIdentifier(true)
              onIdentifierChangeCallback?.(identifierFromName)
            }}
            className={css.idValue}
          />
        </Popover>
        {isIdentifierEditable && !editable ? (
          <Icon
            name="Edit"
            size={12}
            style={{ verticalAlign: 'middle', cursor: 'pointer' }}
            onClick={() => {
              setEditable(true)
            }}
          />
        ) : null}
      </Layout.Horizontal>
      <FormInput.Text
        {...inputGroupProps}
        label={inputLabel}
        name={inputName}
        onChange={e => {
          setCurrentEditField(inputName)
          const name =
            (e.target as HTMLInputElement).value.length > maxInput
              ? formik.values[currentEditField]
              : (e.target as HTMLInputElement).value

          // formik.setFieldValue(inputName, name)
          // isIdentifierEditable && !userModifiedIdentifier && formik.setFieldValue(idName, getIdentifierFromName(name))

          const updatedValues = cloneDeep(formik.values)
          set(updatedValues, inputName, name)
          const identifierFromName = getIdentifierFromName(name)

          if (isIdentifierEditable && !userModifiedIdentifier) {
            set(updatedValues, idName, identifierFromName)
          }

          formik.setValues(updatedValues)

          // onIdentifierChangeCallback needs to be called after the above formik.setValues
          if (isIdentifierEditable && !userModifiedIdentifier) {
            onIdentifierChangeCallback?.(identifierFromName)
          }
        }}
      />
      {!formik.errors[inputName] && formik.errors[idName] ? (
        <FormError name={idName} className={css.idError} errorMessage={get(formik.errors, idName)} />
      ) : null}
      {formik.values[currentEditField]?.length >= maxInput ? (
        <FormError
          name={currentEditField === idName ? idLabel : inputLabel}
          errorMessage={`Limit of ${maxInput} characters is reached for ${
            currentEditField === idName ? idLabel : inputLabel
          }`}
        />
      ) : null}
    </div>
  )
}
