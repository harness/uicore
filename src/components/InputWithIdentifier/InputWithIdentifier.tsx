import React, { useState } from 'react'
import { EditableText, Popover, PopoverInteractionKind } from '@blueprintjs/core'
import { get } from 'lodash-es'
import cx from 'classnames'

import { FormInput, TextProps } from '../FormikForm/FormikForm'
import { Text } from '../Text/Text'
import { Icon } from '../../icons/Icon'
import { Layout } from '../../layouts/Layout'
import { Container } from '../Container/Container'

import textCss from '../TextInput/TextInput.css'
import css from './InputWithIdentifier.css'

export interface InputWithIdentifierProps {
  formik: any
  inputGroupProps?: Omit<TextProps, 'name' | 'label' | 'onChange'>
  /**
   * @default name
   */
  inputName?: string
  /**
   * @default identifier
   */
  idName?: string
  /**
   * @default Name
   */
  inputLabel?: string
  /**
   * @default ID
   */
  idLabel?: string
  /**
   * Should the user be allowed to edit the identifier?
   * @default true
   */
  isIdentifierEditable?: boolean
  /**
   * @default 63
   */
  maxInput?: number
}

// https://harness.atlassian.net/wiki/spaces/CDNG/pages/736200458/Entity+Identifier
export function getIdentifierFromName(str: string): string {
  return str
    .trim()
    .replace(/[^0-9a-zA-Z_$ ]/g, '') // remove special chars except _ and $
    .replace(/ +/g, '_') // replace spaces with _
}

export const InputWithIdentifier: React.FC<InputWithIdentifierProps> = props => {
  const {
    formik,
    inputLabel = 'Name',
    idLabel = 'ID',
    inputName = 'name',
    idName = 'identifier',
    inputGroupProps,
    isIdentifierEditable = true,
    maxInput = 63
  } = props
  const [editable, setEditable] = useState(false)
  const [userModifiedIdentifier, setUserModifiedIdentifier] = useState(false)
  const identifier = get(formik.values, idName) as string
  const [currentEditField, setCurrentEditField] = useState(inputLabel)
  return (
    <div className={cx(css.txtNameContainer, textCss.main)}>
      <Layout.Horizontal className={css.txtIdContainer} spacing="xsmall">
        <Text>{idLabel}:</Text>
        <Popover
          content={<Container padding="small">{identifier}</Container>}
          interactionKind={PopoverInteractionKind.HOVER}
          hoverOpenDelay={500}
          disabled={!identifier || identifier.length < 10}>
          <EditableText
            maxLength={maxInput}
            disabled={!isIdentifierEditable}
            placeholder=""
            minWidth={0}
            value={identifier}
            isEditing={editable}
            onConfirm={() => setEditable(false)}
            onCancel={() => setEditable(false)}
            onEdit={() => setEditable(true)}
            onChange={value => {
              setCurrentEditField(idName)
              formik.setFieldValue(idName, getIdentifierFromName(value))
              setUserModifiedIdentifier(true)
            }}
            className={css.idValue}
          />
        </Popover>
        {isIdentifierEditable && !editable ? (
          <Icon
            name="edit"
            size={12}
            style={{ verticalAlign: 'middle', cursor: 'pointer' }}
            onClick={() => {
              setEditable(true)
            }}
          />
        ) : null}
      </Layout.Horizontal>
      <FormInput.Text
        {...inputGroupProps}
        label={inputLabel}
        name={inputName}
        onChange={e => {
          setCurrentEditField(inputName)
          const name =
            (e.target as HTMLInputElement).value.length > maxInput
              ? formik.values[currentEditField]
              : (e.target as HTMLInputElement).value
          formik.setFieldValue(inputName, name)
          isIdentifierEditable && !userModifiedIdentifier && formik.setFieldValue(idName, getIdentifierFromName(name))
        }}
      />
      {formik.errors[idName] ? (
        <Text font="small" intent="danger" padding={{ bottom: 'medium' }}>
          {get(formik.errors, idName)}
        </Text>
      ) : null}
      {formik.values[currentEditField]?.length >= maxInput ? (
        <Text font="small" intent="danger" padding={{ bottom: 'medium' }}>
          {`Limit of ${maxInput} characters is reached for ${currentEditField === idName ? idLabel : inputLabel}`}
        </Text>
      ) : null}
    </div>
  )
}
