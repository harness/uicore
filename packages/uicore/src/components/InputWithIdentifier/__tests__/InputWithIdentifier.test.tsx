/*
 * Copyright 2024 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { queryByAttribute, render, within } from '@testing-library/react'
import { FormInput } from 'components/FormikForm/FormikForm'
import userEvent from '@testing-library/user-event'
import { renderFormikForm } from 'components/FormikForm/__tests__/FormikForm.test'
import { getIdentifierFromName } from '../InputWithIdentifier'

describe('Test InputWithIdentifier', () => {
  function addNameAndAssertModifiedIdentifier(
    name: string,
    identifier: string,
    input: HTMLInputElement,
    idClass: HTMLInputElement
  ) {
    userEvent.type(input, name)
    expect(input).toHaveValue(name)
    expect(within(idClass).getByText(identifier)).toBeInTheDocument()
    userEvent.clear(input)
    expect(input).toHaveValue('')
  }
  test('allowHyphen false', async () => {
    const { container } = render(
      renderFormikForm(
        <FormInput.InputWithIdentifier
          inputName="name"
          idName="identifier"
          inputLabel="Name"
          idLabel="ID"
          // not passing allowHyphen as its optional
        />
      )
    )
    const input = queryByAttribute('name', container, 'name') as HTMLInputElement
    const idClass = container.querySelector('div.idValue') as HTMLInputElement

    expect(input).toBeInTheDocument()
    expect(idClass).toBeInTheDocument()

    addNameAndAssertModifiedIdentifier('Test Name', 'Test_Name', input, idClass)
    addNameAndAssertModifiedIdentifier('Test Name-123', 'Test_Name123', input, idClass)
    addNameAndAssertModifiedIdentifier('Test Name-123-', 'Test_Name123', input, idClass)
    addNameAndAssertModifiedIdentifier('-Test Name-123-', 'Test_Name123', input, idClass)
    addNameAndAssertModifiedIdentifier('-$Test Name-123-', 'Test_Name123', input, idClass)
    addNameAndAssertModifiedIdentifier('-$Test Name-123-', 'Test_Name123', input, idClass)
    addNameAndAssertModifiedIdentifier('_-$33var1--', '_$33var1', input, idClass)
    addNameAndAssertModifiedIdentifier('_-$33var-abc-1--', '_$33varabc1', input, idClass)
  })

  test('allowHyphen true', async () => {
    const { container } = render(
      renderFormikForm(
        <FormInput.InputWithIdentifier
          inputName="name"
          idName="identifier"
          inputLabel="Name"
          idLabel="ID"
          allowHyphen={true}
        />
      )
    )
    const input = queryByAttribute('name', container, 'name') as HTMLInputElement
    const idClass = container.querySelector('div.idValue') as HTMLInputElement

    expect(input).toBeInTheDocument()
    expect(idClass).toBeInTheDocument()

    addNameAndAssertModifiedIdentifier('Test Name', 'Test_Name', input, idClass)
    addNameAndAssertModifiedIdentifier('Test Name-123', 'Test_Name-123', input, idClass)
    addNameAndAssertModifiedIdentifier('Test Name-123-', 'Test_Name-123-', input, idClass)
    addNameAndAssertModifiedIdentifier('-Test Name-123-', 'Test_Name-123-', input, idClass)
    addNameAndAssertModifiedIdentifier('-$Test Name-123-', 'Test_Name-123-', input, idClass)
    addNameAndAssertModifiedIdentifier('-$Test Name-123-', 'Test_Name-123-', input, idClass)
    addNameAndAssertModifiedIdentifier('_-$33var1--', '_-$33var1--', input, idClass)
    addNameAndAssertModifiedIdentifier('_-$33var-abc-1--', '_-$33var-abc-1--', input, idClass)
  })

  test('test getIdentifierFromName() function', () => {
    expect(getIdentifierFromName('Test Name', true)).toBe('Test_Name')
    expect(getIdentifierFromName('Test Name', false)).toBe('Test_Name')

    expect(getIdentifierFromName('Test Name-123', true)).toBe('Test_Name-123')
    expect(getIdentifierFromName('Test Name-123', false)).toBe('Test_Name123')

    expect(getIdentifierFromName('Test Name-123-', true)).toBe('Test_Name-123-')
    expect(getIdentifierFromName('Test Name-123-', false)).toBe('Test_Name123')

    expect(getIdentifierFromName('-Test Name-123- ', true)).toBe('Test_Name-123-')
    expect(getIdentifierFromName('-$Test Name-123- ', true)).toBe('Test_Name-123-')
    expect(getIdentifierFromName('-$Test Name-123- ', true)).toBe('Test_Name-123-')
    //
    expect(getIdentifierFromName('-Test Name-123- ', false)).toBe('Test_Name123')
    expect(getIdentifierFromName('-$Test Name-123- ', false)).toBe('Test_Name123')
    expect(getIdentifierFromName('-$Test Name-123- ', false)).toBe('Test_Name123')

    expect(getIdentifierFromName('_-$33var1--', true)).toBe('_-$33var1--')
    expect(getIdentifierFromName('_-$33var1--', false)).toBe('_$33var1')

    expect(getIdentifierFromName('_-$33var-abc-1--', true)).toBe('_-$33var-abc-1--')
    expect(getIdentifierFromName('_-$33var-abc-1--', false)).toBe('_$33varabc1')
  })
})
