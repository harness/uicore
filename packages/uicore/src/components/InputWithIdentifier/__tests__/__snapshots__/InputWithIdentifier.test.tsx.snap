// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test InputWithIdentifier allowHyphen false 1`] = `
<div>
  <div
    class="overlaySpinner"
  >
    <form
      action="#"
      class="main"
    >
      <div>
        <div
          class="txtNameContainer main"
        >
          <div
            class="horizontal layout-spacing-xsmall StyledProps--main txtIdContainer"
          >
            <div
              class="withDocsTooltip"
              data-tooltip-id="universalInputNameWithId"
            >
              <p
                class="StyledProps--font StyledProps--main"
              >
                ID
              </p>
            </div>
            <p
              class="StyledProps--font StyledProps--main"
            >
              :
            </p>
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <div
                  class="bp3-editable-text bp3-editable-text-placeholder idValue"
                  tabindex="0"
                >
                  <span
                    class="bp3-editable-text-content"
                    style="height: 1px; line-height: 1px; min-width: 0;"
                  />
                </div>
              </span>
            </span>
            <span
              class="bp3-icon StyledProps--main"
              data-icon="Edit"
              style="vertical-align: middle; cursor: pointer;"
            >
              <svg
                height="12"
                viewBox="0 0 16 16"
                width="12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M11.207.485a.5.5 0 01.354-.146 3.5 3.5 0 013.5 3.5.5.5 0 01-.146.353l-11 11a.5.5 0 01-.354.147h-3a.5.5 0 01-.5-.5v-3a.5.5 0 01.146-.354l11-11zm.553.862L1.061 12.046v2.293h2.293l10.7-10.7a2.5 2.5 0 00-2.294-2.292z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
                <path
                  clip-rule="evenodd"
                  d="M11.89 2.646a.5.5 0 010 .708L8.353 6.889a.5.5 0 01-.708-.707l3.536-3.536a.5.5 0 01.707 0z"
                  fill="currentColor"
                  fill-rule="evenodd"
                  opacity="0.3"
                />
              </svg>
            </span>
          </div>
          <div
            class="bp3-form-group"
            data-id="name-0"
          >
            <label
              class="bp3-label"
              for="name"
            >
              <span
                class="acenter"
                data-tooltip-id="testform_name"
              >
                Name
              </span>
               
              <span
                class="bp3-text-muted"
              />
            </label>
            <div
              class="bp3-form-content"
            >
              <div
                class="bp3-input-group"
              >
                <input
                  autocomplete="off"
                  class="bp3-input"
                  name="name"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <button
          aria-label="Submit"
          class="bp3-button button StyledProps--font StyledProps--main StyledProps--intent-primary with-current-color"
          type="submit"
        >
          <span
            class="bp3-button-text"
          >
            Submit
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
